sdk/storage/azure-storage-blob/azure/storage/blob/_shared/streams.py

```diff
@@ -142,6 +182,69 @@ def tell(self) -> int:
         else:
             raise StructuredMessageError(f"Invalid SMRegion {self._current_region}")
 
+    def seek(self, offset: int, whence: int = SEEK_SET) -> int:
+        if not self.seekable():
+            raise UnsupportedOperation("Inner stream is not seekable.")
+
+        if whence == SEEK_SET:
+            position = offset
+        elif whence == SEEK_CUR:
+            position = self.tell() + offset
+        elif whence == SEEK_END:
+            position = self.message_length + offset
+        else:
+            raise ValueError(f"Invalid value for whence: {whence}")
+
+        if position > self.tell():
+            raise UnsupportedOperation("This stream only supports seeking backwards.")
+
+        # MESSAGE_HEADER
+        if position < self._message_header_length:
+            self._current_region = SMRegion.MESSAGE_HEADER
+            self._current_region_offset = position
+            self._content_offset = 0
+            self._current_segment_number = 0
+        # MESSAGE_FOOTER
+        elif position >= self.message_length - self._message_footer_length:
+            self._current_region = SMRegion.MESSAGE_FOOTER
+            self._current_region_offset = position - (self.message_length - self._message_footer_length)
+            self._content_offset = self.content_length
+            self._current_segment_number = self._num_segments
+        else:
+            # The size of a "full" segment. Fine to use for calculating new segment number and pos
+            full_segment_size = self._segment_header_length + self._segment_size + self._segment_footer_length
+            new_segment_num = 1 + (position - self._message_header_length) // full_segment_size
+            segment_pos = (position - self._message_header_length) % full_segment_size
+            previous_segments_size = (new_segment_num - 1) * self._segment_size
```

Comment:
can this be called something closer to `previous_segments_total_content_size`? It's a lot, but there's a lot of variables going around and this is very clear exactly what it's measuring, especially with the difference between segment content length and segment full length.